lighthouse-core/scripts/generate-bundlephobia-database.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+/* eslint-disable no-console */
+
+// TODO(saavan) These type definitions are incorrect and need to be fixed.
+/** @typedef {{gzip: number, name: string, version: string | number, repository: string}} BundlePhobiaLibrary */
+
+const fs = require('fs');
+const getPackageVersionList = require('bundle-phobia-cli').fetchPackageStats.getPackageVersionList;
+const fetchPackageStats = require('bundle-phobia-cli').fetchPackageStats.fetchPackageStats;
+
+const databasePath = '../audits/byte-efficiency/bundlephobia-database.json';
+
+/** @type {Record<string, string[]>} */
+const suggestionsJSON = require('../audits/byte-efficiency/library-suggestions.js').suggestions;
+/** @type string[] */
+/* eslint-disable-next-line max-len */
+const librarySuggestions = Object.keys(suggestionsJSON).map(s => suggestionsJSON[s].concat(s)).flat();
+
+/** @type {Record<string, Record<'lastScraped', number|string> | Record<'repository', string> | Record<string, BundlePhobiaLibrary>>} */
+let database = {};
+if (fs.existsSync(databasePath)) {
+  database = require(databasePath);
+}
+
+/**
+ * Returns true if this library has been scraped from BundlePhobia in the past week.
```

Comment:
a nice idea, but won't we be running the script at the same time for all the libraries? I guess I'm just struggling to see when they will be different other than when we add new ones.

I'm probably just missing the scenario where we need this caching that you've already thought of :)

Reply:
Yeah it's for when we add new ones, to prevent having to re-scrape everything. While I was making the script, I was testing a lot of diff suggestions, so this was a useful util to speed things up

The other reason is if a specific (or multiple) libraries error'd out while being scraped due to excessive network requests to BundlePhobia, then you'd have to wait a few min to re-run the script but you'd only want those libraries to be re-scraped otherwise you'll run into the same issue. So in summary, to avoid accidental DDOS lol

Reply:
Ah, gotcha 👍 maybe a comment to this effect about rate-limiting then rather than "in the past week" we really care if it's been run in the past like hour 😄

Reply:
Kk 👍